### PR TITLE
Merge Wip/headnode bootstrap error branch to develop

### PR DESCRIPTION
**Pull request by chenwany**
### Description of changes
* Merged feature branch to surface headnode bootstrap failures to develop branch
* The main changes are:
  * Updated the general cfn failure message. When the head node fails bootstrap, we would like to show some general "catch-all" error message if no specific error messages are available. The previous message "Failed to run cfn-init ......" is not customer-friendly since it requires understanding of internal mechanism (cfn-init). We update the error message here in this PR.
  * Add `  "failures": [
    {
      "failureCode": "...",
      "failureReason": "..."
    }
  ]` to describe-cluster output when cluster creation fail
* Included PRs:
  * https://github.com/aws/aws-parallelcluster/pull/4554
  * https://github.com/aws/aws-parallelcluster/pull/4540
  * https://github.com/aws/aws-parallelcluster/pull/4559
  * https://github.com/aws/aws-parallelcluster/pull/4566
  * https://github.com/aws/aws-parallelcluster/pull/4801
  * https://github.com/aws/aws-parallelcluster/pull/4815

### Tests
* Test through manually tests and integration test: https://github.com/aws/aws-parallelcluster/pull/4830
### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/resources/head_node/user_data.sh [hunk 1]
```diff
@@ -126,7 +129,7 @@ if [ "${!custom_cookbook}" != "NONE" ]; then
 fi
 
 # Call CloudFormation
-cfn-init -s ${AWS::StackName} -v -c default -r HeadNodeLaunchTemplate --region ${AWS::Region} || error_exit 'Failed to run cfn-init. If --norollback was specified, check /var/log/cfn-init.log and /var/log/cloud-init-output.log.'
+cfn-init -s ${AWS::StackName} -v -c default -r HeadNodeLaunchTemplate --region ${AWS::Region} || error_exit 'Failed to bootstrap the head node. Please check /var/log/cfn-init.log or /var/log/chef-client.log in the head node, or check the cfn-init.log or chef-client.log in CloudWatch logs. Please refer to https://docs.aws.amazon.com/parallelcluster/latest/ug/troubleshooting-v3.html#troubleshooting-v3-get-logs for more details on ParallelCluster logs.'
```
Comment (davprat) [inline cli/src/pcluster/resources/head_node/user_data.sh:132]: [nitpick] I wonder if this line might not be more readable if we put the failure message in a variable?

File: cli/src/pcluster/resources/head_node/user_data.sh [hunk 2]
```diff
@@ -66,7 +66,10 @@ function error_exit
 {
   # wait logs flush before signaling the failure
   sleep 10
-  reason=$(cat /var/log/parallelcluster/chef_error_msg 2>/dev/null) || reason="$1"
+  # trim the error message because there is a size limit of 4096 bytes for cfn-signal
+  # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cloudformation-limits.html
+  cutoff=$(expr 4096 - $(stat --printf="%s" /tmp/wait_condition_handle.txt))
```
Comment (davprat) [inline cli/src/pcluster/resources/head_node/user_data.sh:71]: I recall that the actual limit (at least in the Cfn console), was something less than 1024?

File: cli/src/pcluster/models/common.py [hunk 3]
```diff
@@ -245,14 +245,20 @@ def _download_s3_objects_with_prefix(self, task_id, destdir):
...
-    """Save CFN stack events into a file."""
+def get_all_stack_events(stack_name: str):
+    """Retrieve all stack events."""
     stack_events = []
     chunk = AWSApi.instance().cfn.get_stack_events(stack_name)
     stack_events.append(chunk["StackEvents"])
     while chunk.get("nextToken"):
         chunk = AWSApi.instance().cfn.get_stack_events(stack_name, next_token=chunk["nextToken"])
         stack_events.append(chunk["StackEvents"])
+    return stack_events
```
Comment (davprat) [inline cli/src/pcluster/models/common.py:256]: [Question] Why do we realize this into a list and not return it as a generator?

File: api/client/src/docs/Failure.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Failure
+
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
+**failure_code** | **str** | Failure code when the cluster stack is in CREATE_FAILED status. | [optional] 
+**failure_reason** | **str** | Failure reason when the cluster stack is in CREATE_FAILED status. | [optional] 
+**any string name** | **bool, date, datetime, dict, float, int, list, str, none_type** | any string name can be used but the value must be the correct type | [optional]
```
Comment (davprat) [inline api/client/src/docs/Failure.md:9]: [Nitpick] Should the description be capitalized?

**Review by davprat — APPROVED** [review]

File: api/client/src/docs/Failure.md (hunk 4 above)
Comment (chenwany) [reply api/client/src/docs/Failure.md:9]: This is auto generated

File: cli/src/pcluster/resources/head_node/user_data.sh (hunk 2 above)
Comment (chenwany) [reply cli/src/pcluster/resources/head_node/user_data.sh:71]: This commit is from the PR from @delongmeng. @delongmeng Could you check the actual limits on it?

 I think we also trim the url here in cookbook to make the size:
https://github.com/aws/aws-parallelcluster-cookbook/pull/1688/commits/e87072fe4ee70865b1ca616751c4073558cfd22d

I am going to create a PR to address the security, once the message is static, I think we don't need to worry about this limit.

File: cli/src/pcluster/resources/head_node/user_data.sh (hunk 1 above)
Comment (chenwany) [reply cli/src/pcluster/resources/head_node/user_data.sh:132]: Thanks for your suggestion! Yeah I think we can put it in variable to improve readability as the message becomes longer now @delongmeng

File: cli/src/pcluster/models/common.py (hunk 3 above)
Comment (chenwany) [reply cli/src/pcluster/models/common.py:256]: The function is used in two places. Forcing it at this level can reduce code to make it in the list in two places

**Pull request merged**